src/resdk/uploader.py

```diff
@@ -0,0 +1,215 @@
+"""The uploader class.
+
+Supports uploading to S3 bucket and to Genialis server.
+"""
+
+import uuid
+from enum import Enum, auto
+from pathlib import Path
+from typing import TYPE_CHECKING, Dict, Optional, Union
+from urllib.parse import urljoin
+
+import boto3
+import botocore
+
+from resdk.aws.session import get_refreshable_boto3_session
+from resdk.constants import CHUNK_SIZE
+from resdk.exceptions import ResolweServerError
+
+if TYPE_CHECKING:
+    from resdk.resolwe import Resolwe
+
+
+class UploadType(Enum):
+    """Supported uploader types."""
+
+    LOCAL = auto()
+    S3 = auto()
+
+    @staticmethod
+    def default():
+        """Return default upload type."""
+        return UploadType.LOCAL
+
+
+class Uploader:
+    """The uploader class.
+
+    It is used to upload the file to the Genialis platform. Is supports
+    uploading files to the server and also directly to the S3 bucket.
+    """
+
+    def __init__(self, resolwe: "Resolwe"):
+        """Store the ."""
+        self.resolwe = resolwe
+        self.upload_methods = {
+            UploadType.LOCAL: self._upload_local,
+            UploadType.S3: self._upload_s3,
+        }
+        # Use the following upload type if one can not be determined.
+        self._default_upload_type = UploadType.LOCAL
+        self._boto_session: Optional[boto3.Session] = None
+        self._upload_config: Optional[Dict] = None
+
+    def invalidate_cache(self):
+        """Remove local cache for upload configuration."""
+        self._upload_config = None
+
+    @property
+    def upload_config(self) -> Dict:
+        """Get the upload configuration.
+
+        If configuration can not be read from the server the default
+        configuration with local upload is returned.
+
+        Use cached version if available.
+        """
+        if self._upload_config is None:
+            try:
+                self._upload_config = self.resolwe.api.upload_config.get()
+            except ResolweServerError:
+                self.resolwe.logger.exception("Upload config could not be retrieved.")
+                self._upload_config = {"type": "LOCAL"}
```

Comment:
Maybe rather `{"type": UploadType.default()}` ?

Reply:
Almost: `UploadType.default().name` to return the compatible config.